android/src/main/java/com/twiliorn/library/PatchedVideoView.java

```diff
@@ -45,6 +64,108 @@ public void run() {
             });
         }
         super.onFrame(frame);
+
+        if (captureThisFrame) {
+            captureThisFrame = false;
+            saveVideoFrame(frame);
+        }
+    }
+
+    /**
+     * Create an NV21Buffer with the same pixel content as the given I420 buffer. Ripped from here: https://stackoverflow.com/a/55158228
+     */
+    private byte[] createNV21Data(VideoFrame.I420Buffer i420Buffer) {
+        final int width = i420Buffer.getWidth();
+        final int height = i420Buffer.getHeight();
+        final int chromaStride = width;
+        final int chromaWidth = (width + 1) / 2;
+        final int chromaHeight = (height + 1) / 2;
+        final int ySize = width * height;
+        final ByteBuffer nv21Buffer = ByteBuffer.allocateDirect(ySize + chromaStride * chromaHeight);
+        // We don't care what the array offset is since we only want an array that is direct.
+        @SuppressWarnings("ByteBufferBackingArray") final byte[] nv21Data = nv21Buffer.array();
+        for (int y = 0; y < height; ++y) {
+            for (int x = 0; x < width; ++x) {
+                final byte yValue = i420Buffer.getDataY().get(y * i420Buffer.getStrideY() + x);
+                nv21Data[y * width + x] = yValue;
+            }
+        }
+        for (int y = 0; y < chromaHeight; ++y) {
+            for (int x = 0; x < chromaWidth; ++x) {
+                final byte uValue = i420Buffer.getDataU().get(y * i420Buffer.getStrideU() + x);
+                final byte vValue = i420Buffer.getDataV().get(y * i420Buffer.getStrideV() + x);
+                nv21Data[ySize + y * chromaStride + 2 * x + 0] = vValue;
+                nv21Data[ySize + y * chromaStride + 2 * x + 1] = uValue;
+            }
+        }
+        return nv21Data;
+    }
+
+    private void sendEvent(String event, WritableMap params) {
+        ((ReactContext) this.getContext()).getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
+                .emit(event, params);
+    }
+
+    private void saveVideoFrame(VideoFrame frame) {
+        frame.retain();
+        Context context = this.getContext();
+        String fileId = UUID.randomUUID().toString();
+        String filename = "rntframe-" + fileId + ".jpeg";
+        VideoFrame.I420Buffer i420Buffer = frame.getBuffer().toI420();
+        final int width = i420Buffer.getWidth();
+        final int height = i420Buffer.getHeight();
+        byte[] nv21Data = createNV21Data(i420Buffer);
+        YuvImage yuvImage = new YuvImage(nv21Data, ImageFormat.NV21, width, height, null);
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        yuvImage.compressToJpeg(new Rect(0, 0, width, height), 100, out);
+        byte[] rawImageBytes = out.toByteArray();
+        byte[] imageBytes = rotateJPEGByteArray(rawImageBytes, 90f);
```

Comment:
❓ Why do you need to rotate 90 degrees always? Is this required for all devices, or does it depend on the device's physical camera orientation? or does it depend on the screen orientation?

Reply:
+1

Reply:
To add here, any image manipulation should be done in a background thread. I don't think we'll get into the space of needing to queue them with the frequency we can make streemshots. Is there a thread limit we need to worry about that you're aware of @garyp?

As I said, I don't think we'll hit the limit, but it is something to consider.

Reply:
The images were being saved to the device rotated so this straightened them out. For reasons I do not know currently

Reply:
I think we will need to understand those reasons before this is ready to merge. I suspect the rotation is not universal but will depend on the device's characteristics, which means 90 degrees won't always be the right number.

Reply:
Yeah, true - that's where a thread limit came to mind. I haven't worked in concurrency on iOS in a while, but I recall some gotchas to be aware of. After seeing your comments, I shared some articles I found to remind myself of details with Chris. We'll explore those as we refine the solution.

Reply:
Ok I think I see why the rotation is needed although still not sure whats going on.

I got this message when the CameraCapturer is created from this log at this [LOC](https://github.com/streem/react-native-twilio-video-webrtc/blob/main/android/src/main/java/com/twiliorn/library/CustomTwilioVideoView.java#L258)
"Creating new CameraCapturer, cameraId: Camera 0, Facing back, Orientation 90"

I think its that orientation of 90 from the camera capturer which is why I needed to apply the rotation before saving images. 

Idk why camera capturer orientation is 90 but thats my first clue

Reply:
We were able to figure out that the `VideoFrame` provides its captured orientation and use that to rotate the image bytes dynamically instead of hardcoding